src/core_deku/ticket_table.ml

```diff
@@ -0,0 +1,126 @@
+open Helpers
+
+module Tickets = struct
+  module Ticket_map = Map.Make_with_yojson (Ticket_id)
+
+  type t = Amount.t Ticket_map.t [@@deriving yojson]
+
+  let get_balance t ~ticket = Ticket_map.find_opt ticket t
+
+  let of_seq = Ticket_map.of_seq
+
+  let singleton ~ticket ~amount =
+    Ticket_map.of_seq (Seq.return (ticket, amount))
+
+  let merge t ~ticket ~amount =
+    let merger option ~amount =
+      let value = Option.fold ~none:amount ~some:Amount.(( + ) amount) option in
+      value in
+    Ticket_map.update ticket
+      (fun x ->
+        let merged = merger ~amount x in
+        Option.some merged)
+      t
+
+  let get_and_remove t ~ticket ~to_take =
+    let%ok amount =
+      Ticket_map.find_opt ticket t |> Option.to_result ~none:`Insufficient_funds
+    in
+    let%assert () =
+      ( `Insufficient_funds,
+        let compared = Amount.compare amount to_take in
+        compared >= 0 ) in
+    let map =
+      Ticket_map.update ticket
+        (function
+          | Some amount ->
+            if Amount.equal amount to_take then
+              None
+            else
+              Some Amount.(amount - to_take)
+          | None -> assert false)
+        t in
+    Ok map
+
+  let subtract t ~ticket ~to_withdraw =
+    let%ok map = get_and_remove t ~ticket ~to_take:to_withdraw in
+    Ok map
+
+  let fetch t tickets =
+    let fetched =
+      List.fold_left_ok
+        (fun (lst, m) (ticket, amount) ->
+          let%ok m = get_and_remove m ~ticket ~to_take:amount in
+          let lst = Seq.cons (ticket, amount) lst in
+          Ok (lst, m))
+        (Seq.empty, t) tickets in
+    fetched
+
+  let fold = Ticket_map.fold
+
+  let empty = Ticket_map.empty
+
+  let is_empty = Ticket_map.is_empty
+end
+
+module Table = Map.Make_with_yojson (Address)
+
+type t = Tickets.t Table.t [@@deriving yojson]
+
+let empty = Table.empty
+
+let balance t ~sender ~ticket =
+  let%some map = Table.find_opt sender t in
+  Tickets.get_balance ~ticket map
+
+let update_tickets t ~sender ~tickets =
+  if Seq.is_empty tickets then
```

Comment:
This function seems weird to me. 

Either you're depositing tickets, in which case just use `deposit` directly.

Or you're just garbage collecting `Table.remove`. If that's necessary, why not do it automatically when you withdraw all the tickets? If it's not necessary, why do it at all?

Reply:
when you transfer all of your tickets from one deku account to another one, reference to the old one inside the ticket table should be deleted, to not keep old empty entries in the map

Reply:
I don't get it - isn't this addressed by my second bullet point?

But I see the code is removed. Resolving.